lib/db2.js

```diff
@@ -442,80 +443,66 @@ DB2.prototype.update = function(model, where, data, options, cb) {
 DB2.prototype.updateOrCreate = DB2.prototype.save =
   function(model, data, options, callback) {
     var self = this;
-    var fields = self.buildFields(model, data);
     var id = self.idName(model);
-    var sql = new ParameterizedSQL('MERGE INTO ' +
-                                   self.schema.toUpperCase() + '.' +
-                                   self.tableEscaped(model));
-    var columnValues = fields.columnValues;
-    var fieldNames = fields.names;
-    var setValues = [];
-
-    var definition = self.getModelDefinition(model);
-
-    if (fieldNames.length) {
-      sql.merge('AS MT(');
-
-      Object.keys(definition.properties).forEach(function(prop) {
-        setValues.push(new ParameterizedSQL(self.columnEscaped(model, prop)));
-      });
-
-      var columns = ParameterizedSQL.join(setValues, ',');
-      sql.merge(columns);
-      sql.merge(')');
-      sql.merge('USING (SELECT * FROM TABLE( VALUES(');
-
-      setValues = [];
-      for (var i = 0, n = fields.names.length; i < n; i++) {
-        setValues.push(new ParameterizedSQL('CAST (' + columnValues[i].sql +
-          ' AS ' +
-          self.buildColumnType(fields.properties[i]) + ')',
-          columnValues[i].params));
-
-        if (i < (n - 1))
-          setValues[i].sql = setValues[i].sql + ',';
-      }
-
-      sql.merge(setValues);
-      sql.merge('))) AS VT(' + fieldNames.join(',') + ')');
-      sql.merge('ON');
-      sql.merge('(MT.\"' + id + '\" = VT.\"' + id + '\")');
-      sql.merge('WHEN NOT MATCHED THEN INSERT (' +
-      fieldNames.join(',') + ')');
-
-      var values = ParameterizedSQL.join(columnValues, ',');
-      values.sql = 'VALUES(' + values.sql + ')';
-      sql.merge(values);
-    } else {
-      sql.merge(self.buildInsertDefaultValues(model, data, options));
-    }
-
-    sql.merge('WHEN MATCHED THEN UPDATE SET');
-
-    setValues = [];
-    for (i = 0, n = fields.names.length; i < n; i++) {
-      if (!fields.properties[i].id) {
-        setValues.push(new ParameterizedSQL(fields.names[i] + '=' +
-          '(' + columnValues[i].sql + ')', columnValues[i].params));
+    var stmt;
+    var tableName = self.tableEscaped(model);
+    var idData = {};
+    var countData;
+    var isInsert = false;
+    var meta = {};
+    var connection;
+    var currentTx;
+
+    process.nextTick(function() {
+      if (options.transaction) {
+        connection = options.transaction.connection;
+      } else {
+        try {
+          connection = Driver.openSync(self.connStr);
```

Comment:
Not really. Other DBs such as MySQL uses async calls to enforce the transaction. Here is the pseudo code:

```
var self = this;
this.beginTransaction(function(err, tx) {
  if (err) return callback(err);
  var options = { transaction: tx };
  self.executeSQL(sql, params, options, function(err, ...) {
  }
} 
```

The key is that when a TX begins, a connection is reserved for the TX. It will be passed via `options` for further SQL calls within the same TX until the TX is committed or rolled back.
```